cola/widgets/toolbar.py

```diff
@@ -0,0 +1,670 @@
+        QtWidgets.QListWidget.__init__(self, parent)
+
+        self.setAcceptDrops(True)
+        self.setSelectionMode(self.SingleSelection)
+        self.setDragEnabled(True)
+        self.setDropIndicatorShown(True)
+
+        self._mixin = self.Mixin(self, QtWidgets.QListWidget)
+
+    def dragEnterEvent(self, event):
+        return self._mixin.dragEnterEvent(event)
+
+    def dragMoveEvent(self, event):
+        return self._mixin.dragMoveEvent(event)
+
+    def dropEvent(self, event):
+        return self._mixin.dropEvent(event)
+
+    def add_separator(self, title, data):
+        item = QtWidgets.QListWidgetItem()
+        item.setText(title)
+        item.setData(Qt.UserRole, data)
+
+        self.addItem(item)
+
+    def add_item(self, title, data, icon_text=None):
+        item = QtWidgets.QListWidgetItem()
+        item.setText(N_(title))
+        item.setData(Qt.UserRole, data)
+
+        if icon_text is not None:
+            icon = getattr(icons, icon_text, None)
+            item.setIcon(icon())
+
+        self.addItem(item)
+
+    def get_items(self):
+        return self._mixin.get_items()
+
+
+class ToolbarTreeWidget(standard.TreeView):
+    def __init__(self, parent):
+        standard.TreeView.__init__(self, parent)
+
+        self.setDragEnabled(True)
+        self.setDragDropMode(QtWidgets.QAbstractItemView.DragOnly)
+        self.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)
+        self.setDropIndicatorShown(True)
+        self.setRootIsDecorated(True)
+        self.setHeaderHidden(True)
+        self.setAlternatingRowColors(False)
+        self.setSortingEnabled(False)
+
+        self.setModel(QtGui.QStandardItemModel())
+
+    def create_item(self, name, data):
+        item = QtGui.QStandardItem()
+
+        item.setEditable(False)
+        item.setDragEnabled(True)
+        item.setText(N_(name))
+        item.setToolTip(N_(name))
+        item.setData(data, Qt.UserRole)
+
+        return item
+
+    def insert_top(self, parent_id, title):
+        item = self.create_item(title, parent_id)
+        item.setFlags(Qt.ItemIsSelectable | Qt.ItemIsEnabled)
+
+        self.model().insertRow(0, item)
+        self.model().sort(0)
+
+        return item
+
+    def insert_child(self, parent_id, child_id, title, icon_text=None):
+        data = {'parent_id': parent_id, 'child_id': child_id}
+        item = self.create_item(title, data)
+
+        if icon_text is not None:
+            icon = getattr(icons, icon_text, None)
+            item.setIcon(icon())
+
+        return item
+
+    def update_top_item(self, parent_title, child_id, title, icon):
+        items = self.model().findItems(N_(parent_title), Qt.MatchExactly)
+
+        if len(items) > 0:
+            top = items[0]
+            child = self.insert_child(top.data(Qt.UserRole), child_id, title,
+                                      icon)
+            top.appendRow(child)
+            top.sortChildren(0, Qt.AscendingOrder)
+
+
+MAIN_ACTIONS = {
```

Comment:
It would be nice if there was an action that ran the `Edit` command, so that the selected files are opened in the editor (same as the Ctrl+e hotkey).  I went looking for it and didn't see an entry for it, though that's something I use pretty often (I might be an atypical user).

Another notable missing item was "Difftool" from the Diff section (same as the Ctrl-d hotkey) so that we can diff the selected files.

I would also recommend making it so that the "Configure toolbar" dialog remembers its window geometry.  I keep having to resizing it everytime I open it so reusing the apply_state() code can do that for us so that it'll remember its size and position.